Empty intersection in HPolygon's addconstraints! resp. isredundant

The intersection of the lines corresponding to two-dimensional constraints can be empty due to numerical issues. Instead of raising an exception, we should add the constraint, making the polygon empty.
In the example below, constraints `Hs[2]` and `Hs[3]` are parallel, and so are `Hs[1]` and `H`. Here the intersection is nonempty.

```julia
julia> Hs = [HalfSpace([0.05125918978413134, 0.0], 0.956012965730266),
 HalfSpace([-0.0363879863894044, 0.3036279569180349], -0.370213880061176),
 HalfSpace([0.03529692609786872, -0.2945239519959644], 0.36249837124872736)];

julia> H = HalfSpace([-0.012707966980287463, 3.809113859067846e-15], -0.2348665397215645);

julia> addconstraint!(Hs, H)
ERROR: AssertionError: $(Expr(:escape, :(cap isa Singleton)))
Stacktrace:
 [1] macro expansion at /.julia/dev/LazySets/src/Assertions/Assertions.jl:23 [inlined]
 [2] isredundant(::HalfSpace{Float64,Array{Float64,1}}, ::HalfSpace{Float64,Array{Float64,1}}, ::HalfSpace{Float64,Array{Float64,1}}) at /.julia/dev/LazySets/src/Interfaces/AbstractHPolygon.jl:320
 [3] #addconstraint!#315(::Bool, ::Bool, ::typeof(addconstraint!), ::Array{HalfSpace{Float64,Array{Float64,1}},1}, ::HalfSpace{Float64,Array{Float64,1}}) at /.julia/dev/LazySets/src/Interfaces/AbstractHPolygon.jl:460
 [4] addconstraint!(::Array{HalfSpace{Float64,Array{Float64,1}},1}, ::HalfSpace{Float64,Array{Float64,1}}) at /.julia/dev/LazySets/src/Interfaces/AbstractHPolygon.jl:436
 [5] top-level scope at REPL[31]:1
 [6] eval(::Module, ::Any) at ./boot.jl:330
 [7] eval_user_input(::Any, ::REPL.REPLBackend) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.2/REPL/src/REPL.jl:86
 [8] run_backend(::REPL.REPLBackend) at /.julia/packages/Revise/Pcs5V/src/Revise.jl:1073
 [9] top-level scope at none:0
```

--- a/src/Arrays/Arrays.jl
+++ b/src/Arrays/Arrays.jl
@@ -9,6 +9,7 @@ using LinearAlgebra, SparseArrays
 using ..Assertions: @assert, activate_assertions
 # activate assertions by default
 activate_assertions(Arrays)
+using ..LazySets: _geq
 
 include("matrix_operations.jl")
 include("vector_operations.jl")

--- a/src/Arrays/vector_operations.jl
+++ b/src/Arrays/vector_operations.jl
@@ -345,11 +345,11 @@ counter-clockwise) with respect to the center `O`.
 @inline function is_right_turn(O::AbstractVector{N},
                                u::AbstractVector{N},
                                v::AbstractVector{N}) where {N<:Real}
-    return right_turn(O, u, v) >= zero(N)
+    return _geq(right_turn(O, u, v), zero(N))
 end
 
 # version for O = origin
 @inline function is_right_turn(u::AbstractVector{N},
                                v::AbstractVector{N}) where {N<:Real}
-    return right_turn(u, v) >= zero(N)
+    return _geq(right_turn(u, v), zero(N))
 end

--- a/src/LazySets.jl
+++ b/src/LazySets.jl
@@ -25,6 +25,11 @@ import .Assertions: activate_assertions, deactivate_assertions
 # activate assertions by default
 activate_assertions(LazySets)
 
+# =====================
+# Numeric approximation
+# =====================
+include("Utils/comparisons.jl")
+
 # =======================
 # Arrays auxiliary module
 # =======================
@@ -35,7 +40,6 @@ using .Arrays
 # Auxiliary functions
 # ===================
 include("Utils/helper_functions.jl")
-include("Utils/comparisons.jl")
 include("Utils/macros.jl")
 
 # ==================

--- a/src/Sets/VPolygon.jl
+++ b/src/Sets/VPolygon.jl
@@ -397,12 +397,7 @@ julia> [4.5, 3.1] ∈ P
 false
 julia> [4.5, 3.0] ∈ P
 true
-julia> [4.4, 3.4] ∈ P  #  point lies on the edge -> floating-point error
-false
-julia> P = VPolygon([[2//1, 3//1], [3//1, 1//1], [5//1, 1//1], [4//1, 5//1]];
-                     apply_convex_hull=false);
-
-julia> [44//10, 34//10] ∈ P  #  with rational numbers the answer is correct
+julia> [4.4, 3.4] ∈ P  #  point lies on the edge
 true
 ```
 """

--- a/test/unit_Polygon.jl
+++ b/test/unit_Polygon.jl
@@ -483,6 +483,14 @@ for N in [Float64]
     I3 = Interval(N(4), N(5))
     @test !is_intersection_empty(I1 × I2 , p3)
     @test is_intersection_empty(I1 × I3 , p3)
+
+    # redundancy with almost-parallel constraints (numeric issues, #2102)
+    Hs = [HalfSpace([0.05125918978413134, 0.0], 0.956012965730266),
+          HalfSpace([-0.0363879863894044, 0.3036279569180349], -0.370213880061176),
+          HalfSpace([0.03529692609786872, -0.2945239519959644], 0.36249837124872736)]
+    H = HalfSpace([-0.012707966980287463, 3.809113859067846e-15], -0.2348665397215645)
+    addconstraint!(Hs, H)
+    @test length(Hs) == 4
 end
 
 # default Float64 constructors

